data/org.freedesktop.portal.CaptureInput.xml

```diff
@@ -0,0 +1,406 @@
+        @options: Vardict with optional further information
+        @fd: A file descriptor to an active EIS implementation that can be passed to a passive libei context
+
+        Set up the connection to an active EIS implementation. Once input capturing starts,
+        input events are sent via the EI protocol between the compositor and the application.
+        This call must be invoked before org.freedesktop.portal.CaptureInput.Enable().
+
+        A session only needs to set this up once, the EIS implementation is not affected by
+        calls to Disable() and org.freedesktop.portal.CaptureInput.Enable() -
+        the same context can be re-used until the session is closed.
+    -->
+    <method name="ConnectToEIS">
+      <arg type="o" name="session_handle" direction="in"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="h" name="fd" direction="out"/>
+    </method>
+
+    <!--
+        Disabled:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @options: Vardict with optional further information
+
+        The Disabled signal is emitted when the application will no longer
+        receive captured input. Any currently ongoing capturing is terminated
+        immediately and pointer barriers or other means of starting input
+        capturing will no longer trigger.
+
+        Applications must call org.freedesktop.portal.CaptureInput.Enable() to
+        request future input capturing for this session.
+
+        This signal may be sent when the application is already in a suspended
+        state.
+    -->
+    <signal name="Disabled">
+      <arg type="o" name="session_handle" direction="out"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+    </signal>
+
+    <!--
+        RegionsChanged:
+        @session_handle: Object path for the #org.freedesktop.portal.Session object
+        @options: Vardict with optional further information
+
+        The RegionsChanged signal is emitted when one or more of the regions
+        available **to this session** change. An application should immediately call
+        org.freedesktop.portal.CaptureInput.GetRegions() to update its state of the regions.
+    -->
+    <signal name="RegionsChanged">
+      <arg type="o" name="session_handle" direction="out"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+    </signal>
+
+    <!--
+        SupportedCapabilities:
+
+        A bitmask of supported capabilities. This list is constant, it is not the list of
+        capabilities currently available but rather which capabilies are
+        implemented by the portal.
+
+        Applications must ignore unknown capabilities.
+
+        Currently defined types are:
+
+        <simplelist>
+          <member>1: RELATIVE_POINTER</member>
+          <member>2: ABSOLUTE_POINTER</member>
+          <member>4: KEYBOARD</member>
+          <member>8: TOUCH</member>
```

Comment:
For triggers, we have DBus methods (currently `SetPointerBarrier` but I can envision at least a `ActivateNow` type too). Those have return values, so I think it may be good enough to just error out from those and let the application deal with it. If in the future we notice that applications need to know ahead of time whether they can set up a barrier, we can add those, separate to the input capabilities. But right now I think they're not needed - doubly so because the capturing triggers will be very few, so it's likely all implementation will support all of them.

Reply:
I'm thinking of when the portal backend should present a dialog to the user. At this point, it should be known what the application wants, both in terms of possilbe triggers and the type of input devices. Doing it at `Enable()` is probably too late, since `ConnectToEIS()` would be called prior to that, which wouldn't be possible since it's not known what kind of permissions there will be. Doing it at `SetPointerBarrier()` wouldn't be good either, since that means adding another delayed trigger would require yet another dialog. My point is that at some point before any `ConnectToEIS()` or `Enable()`, the whole intention of the application must be known, so that a complete picture of its needs can be presented to the user.

Reply:
Yep, we don't disagree. I'm just suggesting to leave this off for now and work it into the various `options a{sv}` and `result a{sv}` if we end up having a specific use-case for it. Probably at `CreateSession()` time, on the assumption that if we allow an application to set up pointer barrier and it ends up setting up the wrong one that's a bug rather than a sandbox violation.

As for "when to present the dialog", I'd say at `CreateSession` time. That currently already has the capabilities field so "An application wants to capture pointer input" is possible there. `SetPointerBarriers` uses a Request and has a return value for failed barriers, so an extra dialog could be inserted here, if that's really needed.

I think (and that needs to be spelled out in the docs) that the EIS implementation needs integrate with the `impl.portal` correctly anyway, because `ConnectToEis()` is only called once but if regions change and/or the client set up invalid pointer barriers later, the fd is still there even though the client must no longer receive events.